src/Microsoft.ApplicationInsights/Managed/Shared/Extensibility/Implementation/Tracing/HeartbeatProvider.cs

```diff
@@ -0,0 +1,400 @@
+        /// </summary>
+        private ConcurrentDictionary<string, HeartbeatPropertyPayload> extendPayloadItems;
+
+        private bool disposedValue = false; // To detect redundant calls to dispose
+        private TimeSpan heartbeatInterval; // time between heartbeats emitted specified in milliseconds
+        private TelemetryClient telemetryClient; // client to use in sending our heartbeat
+        private volatile bool isEnabled; // no need for locks or volatile here, we can skip/add a beat if the module is disabled between heartbeats
+
+        public HeartbeatProvider()
+        {
+            this.disabledDefaultFields = null;
+            this.heartbeatInterval = TimeSpan.FromMilliseconds(DefaultHeartbeatIntervalMs);
+            this.extendPayloadItems = new ConcurrentDictionary<string, HeartbeatPropertyPayload>(StringComparer.OrdinalIgnoreCase);
+            this.sdkPayloadItems = new ConcurrentDictionary<string, HeartbeatPropertyPayload>(StringComparer.OrdinalIgnoreCase);
+            this.heartbeatsSent = 0; // count up from construction time
+            this.isEnabled = false; // wait until Initialize is called before this means anything
+        }
+
+        /// <summary>
+        /// Gets or sets the currently defined interval between heartbeats
+        /// </summary>
+        public TimeSpan Interval
+        {
+            get => this.heartbeatInterval;
+            set
+            {
+                if (value == null || value.TotalMilliseconds <= 0)
+                {
+                    this.heartbeatInterval = TimeSpan.FromMilliseconds(DefaultHeartbeatIntervalMs);
+                }
+                else
+                {
+                    this.heartbeatInterval = value;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets the currently defined instrumentation key to send heartbeat telemetry items to.
+        /// 
+        /// Note that if the heartbeat provider has not been initialized yet, this key would get reset to 
+        /// whatever the telemetry configuration the heartbeat provider is initialized with.
+        /// </summary>
+        public string InstrumentationKey
+        {
+            get => this.telemetryClient?.InstrumentationKey;
+            set
+            {
+                if (this.telemetryClient != null)
+                {
+                    this.telemetryClient.InstrumentationKey = value;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether or not heartbeats are enabled
+        /// </summary>
+        public bool IsEnabled
+        {
+            get => this.isEnabled;
+            set
+            {
+                if (!this.isEnabled && value)
+                {
+                    // we need to start calling the timer again
+                    this.HeartbeatTimer.Change(this.heartbeatInterval, this.heartbeatInterval);
```

Comment:
why are you using private field directly instead of property?

Reply:
Oversight. Fixed.